### PR TITLE
KEA-140 As an admin, I want to upload/update information about announcements

### PR DESCRIPTION
**Pull request by jiahn326**
KEA-140 As an admin, I want to upload/update information about announcements

### CODE REVIEW

File: backend/tests/api/announcement/index.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        adminUser = await userFactory.createUser(strapi, "admin");
+    })
+
+    const constructAnnouncement = (id) => {
+        return {
+            data: {
+                "Title": "Test Announcement " + id,
+                "Poster": null,
+                "Description": "Announcement Description " + id,
+                // "HostingGroup": "Announcement HostingGroup" + id,
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:20]: Why is it commented out? Do we need to tag Hosting Group?

File: backend/tests/api/announcement/index.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            .post("/api/announcements")
+            .set("accept", "application/json")
+            .set("Authorization", `Bearer ${await jwt(adminUser.id)}`)
+            .send(constructAnnouncement(1))
+            .expect("Content-Type", /json/)
+            .expect(200)
+            .expect((data) => {
+                data = data.body.data;
+                expect(data.attributes.Title).toBe(announcement.data.Title);
+                // expected "null", but received "undefined"
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:60]: Is this comment needed? If this is a question for us, make a review comment. If it's an unnecessary comment, it should be deleted.

File: backend/tests/api/announcement/index.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                // expect(data.attributes.Poster).toBe(announcement.data.Poster);
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:61]: Same here

File: backend/tests/api/announcement/index.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                expect(data.body.data.length).toBe(1);
+            })
+    })
+
+    // FindOne
+    it("[FindOne] Public user should find an announcement", async () => {
+        const id = 1;
+        const announcement = constructAnnouncement(id);
+        await request(strapi.server.httpServer)
+            .get("/api/announcements/" + 1)
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:107]: Please use id.

File: backend/tests/api/announcement/index.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                expect(data.attributes.Private).toBe(announcement.data.Private);
+                expect(data.attributes.locale).toBe(announcement.data.locale);
+            })
+    })
+
+    it("[FindOne] Authenticated user should find an announcement", async () => {
+        const id = 1;
+        const announcement = constructAnnouncement(id);
+        await request(strapi.server.httpServer)
+            .get("/api/announcements/" + 1)
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:124]: Same here

File: backend/tests/api/announcement/index.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                expect(data.attributes.Private).toBe(announcement.data.Private);
+                expect(data.attributes.locale).toBe(announcement.data.locale);
+            })
+    })
+
+    it("[FindOne] Admin user should find an announcement", async () => {
+        const id = 1;
+        const announcement = constructAnnouncement(id);
+        await request(strapi.server.httpServer)
+            .get("/api/announcements/" + 1)
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:142]: Same here

File: backend/src/api/announcement/content-types/announcement/schema.json
Comment (doolri1276) [inline backend/src/api/announcement/content-types/announcement/schema.json:1]: Line 37. Do you think we need localization for Poster?

**Review by doolri1276 — CHANGES_REQUESTED** [review]

File: backend/tests/api/announcement/index.js (hunk 2 above)
Comment (jiahn326) [reply backend/tests/api/announcement/index.js:60]: okay sounds good.

File: backend/tests/api/announcement/index.js (hunk 3 above)
Comment (jiahn326) [reply backend/tests/api/announcement/index.js:61]: got it👍

File: backend/config/sync/user-role.admin.json
Comment (doolri1276) [inline backend/config/sync/user-role.admin.json:1]: Please rebase
Screenshot 2024-03-04 at 7 37 51 PM

**Review by doolri1276 — CHANGES_REQUESTED** [review]

File: backend/tests/api/announcement/index.js (hunk 1 above)
Comment (jiahn326) [reply backend/tests/api/announcement/index.js:20]: I tried to test it, but it did not work because I need to set Hosting group beforehand. I can remove that line!

File: backend/tests/api/announcement/index.js (hunk 4 above)
Comment (jiahn326) [reply backend/tests/api/announcement/index.js:107]: gotcha!

File: backend/src/api/announcement/content-types/announcement/schema.json
Comment (jiahn326) [reply backend/src/api/announcement/content-types/announcement/schema.json:1]: yeah I think we need the default photo(church logo)

Comment (doolri1276) [reply backend/src/api/announcement/content-types/announcement/schema.json:1]: Localization for posters means that the poster can have the Kor version and the Eng version. Do you think we need localization for the posters? 

There is no answer. Throwing a question for you to think about it. :)

Comment (jiahn326) [reply backend/src/api/announcement/content-types/announcement/schema.json:1]: Oh I see! Uhhh I don't think we need it.

Comment (doolri1276) [reply backend/src/api/announcement/content-types/announcement/schema.json:1]: If you don't think we need it, could you turn off localization for poster?

File: backend/tests/api/announcement/index.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+describe("Announcement Test", () => {
+    let authenticatedUser;
+    let adminUser;
+
+    beforeAll(async () => {
+        authenticatedUser = await userFactory.createUser(strapi, "authenticated");
+        adminUser = await userFactory.createUser(strapi, "admin");
+    })
+
+    const constructAnnouncement = (id) => {
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:14]: Could you move this function to constructor.js?

File: backend/tests/api/announcement/index.js [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            .expect(200)
+            .then((data) => {
+                data = data.body.data;
+                expect(data.attributes.Title).toBe(announcement.data.Title);
+                expect(data.attributes.Description).toBe(announcement.data.Description);
+                expect(data.attributes.Private).toBe(announcement.data.Private);
+                expect(data.attributes.locale).toBe(announcement.data.locale);
+            })
+    })
+
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:194]: Please delete the additional line. There are two empty lines.

File: backend/tests/api/announcement/index.js [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    beforeAll(async () => {
+        authenticatedUser = await userFactory.createUser(strapi, "authenticated");
+        adminUser = await userFactory.createUser(strapi, "admin");
+    })
+
+    const constructAnnouncement = (id) => {
+        return {
+            data: {
+                "Title": "Test Announcement " + id,
```
Comment (doolri1276) [inline backend/tests/api/announcement/index.js:17]: Can you provide HostingGroup with null?

File: backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json
Comment (doolri1276) [inline backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json:1]: This file is in gitignore as in this file is no longer managed by git. Please remove this file from git.

File: backend/src/extensions/documentation/public/index.html
Comment (doolri1276) [inline backend/src/extensions/documentation/public/index.html:1]: This file is in gitignore as in this file is no longer managed by git. Please remove this file from git.

File: backend/config/sync/user-role.admin.json
Comment (doolri1276) [inline backend/config/sync/user-role.admin.json:1]: ** This comment is not related to this file.

Please work on the view of the Announcement.
Screenshot 2024-03-17 at 9 44 32 PM

Comment (doolri1276) [inline backend/config/sync/user-role.admin.json:1]: ** This comment is not related to this file.
Please work on the view of the Announcement listing.
Screenshot 2024-03-17 at 9 46 00 PM

File: backend/types/generated/contentTypes.d.ts
Comment (doolri1276) [inline backend/types/generated/contentTypes.d.ts:1]: From the admin's viewpoint, do you think Hosting Group should be required or not required?

File: backend/config/sync/user-role.admin.json
Comment (doolri1276) [reply backend/config/sync/user-role.admin.json:1]: @jiahn326 @calvinyoon Rebased successfully. Great work both of you! 🙌

File: backend/types/generated/contentTypes.d.ts
Comment (jiahn326) [reply backend/types/generated/contentTypes.d.ts:1]: I think we should not required for Hosting Group. Because some announcements can be uncertain to set Hosting Group.

File: backend/config/sync/user-role.admin.json
Comment (jiahn326) [reply backend/config/sync/user-role.admin.json:1]: I put ID, Title, Poster, Link. Since we decided to change description to Rich Text, description doesn't show on the view. Are there anything I should show on the view?

Comment (jiahn326) [reply backend/config/sync/user-role.admin.json:1]: Also, StartDate is showing weirdly - so I did not put it on the view.

File: backend/src/extensions/documentation/public/index.html
Comment (doolri1276) [reply backend/src/extensions/documentation/public/index.html:1]: The file is still there.

File: backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json
Comment (doolri1276) [reply backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json:1]: The file is still there.

File: backend/types/generated/contentTypes.d.ts
Comment (doolri1276) [reply backend/types/generated/contentTypes.d.ts:1]: Good point! You are right, some announcements won't be related to any groups.

Could we add a description that they need to set hosting group if it is related to any groups? I want to prevent admins not doing that and asking Hyunjin to do that all the time.

File: backend/config/sync/user-role.admin.json
Comment (jiahn326) [reply backend/config/sync/user-role.admin.json:1]: Do we need Youtube URL field?

File: backend/types/generated/contentTypes.d.ts
Comment (jiahn326) [reply backend/types/generated/contentTypes.d.ts:1]: image
do you think this is aggressive?

Comment (jiahn326) [reply backend/types/generated/contentTypes.d.ts:1]: image

Is it look good?

File: backend/config/sync/admin-role.strapi-super-admin.json [hunk 10]
```diff
@@ -77,11 +85,15 @@
...
           "Title",
           "Poster",
-          "Description",
           "HostingGroup",
           "Private",
           "Link.Label",
-          "Link.URL"
+          "Link.URL",
+          "Description",
+          "StartDate.Date",
```
[the remark this answers is not in the document]
Comment (jiahn326) [reply backend/config/sync/admin-role.strapi-super-admin.json]: Removed it.

File: backend/config/sync/admin-role.admin-lrh30qnl.json
[the remark this answers is not in the document]
Comment (jiahn326) [reply backend/config/sync/admin-role.admin-lrh30qnl.json:1]: Yeah I can do that.

1. Made Poster multiple
2. I created YoutubeURL and put it next to the Poster in the view

I did not put any for RegExp pattern. Should we leave it blank for now?

File: backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json
Comment (doolri1276) [reply backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json:1]: @jiahn326 This file is still here.

File: backend/src/extensions/documentation/public/index.html
Comment (doolri1276) [reply backend/src/extensions/documentation/public/index.html:1]: @jiahn326 This file is still there.

File: backend/types/generated/contentTypes.d.ts
Comment (doolri1276) [reply backend/types/generated/contentTypes.d.ts:1]: Looks great!

File: backend/config/sync/admin-role.admin-lrh30qnl.json
[the remark this answers is not in the document]
Comment (jiahn326) [reply backend/config/sync/admin-role.admin-lrh30qnl.json:1]: Done :)

File: backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json
Comment (doolri1276) [reply backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json:1]: Still there. Please self-check on "Files changed" to make sure that this file is no longer there.

File: backend/src/extensions/documentation/public/index.html
Comment (doolri1276) [reply backend/src/extensions/documentation/public/index.html:1]: Still there. Please self-check on "Files changed" to make sure that this file is no longer there.

File: backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json
Comment (jiahn326) [reply backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json:1]: Finally got 'This file was deleted.'!

File: backend/src/extensions/documentation/public/index.html
Comment (jiahn326) [reply backend/src/extensions/documentation/public/index.html:1]: Finally got 'This file was deleted.'!

File: backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json
Comment (doolri1276) [reply backend/src/extensions/documentation/documentation/1.0.0/full_documentation.json:1]: Yayyyy!

File: backend/config/sync/user-role.admin.json
Comment (doolri1276) [reply backend/config/sync/user-role.admin.json:1]: Is this applied?

File: backend/config/sync/admin-role.strapi-super-admin.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (doolri1276) [reply backend/config/sync/admin-role.strapi-super-admin.json]: Great :)

File: backend/config/sync/admin-role.strapi-super-admin.json [hunk 11]
```diff
@@ -765,10 +768,6 @@
           "Status",
           "Memo",
           "DateCompleted"
-        ],
```
Comment (doolri1276) [inline backend/config/sync/admin-role.strapi-super-admin.json:768]: Please revert all the changes about support

File: backend/config/sync/user-role.admin.json [hunk 12]
```diff
@@ -36,6 +48,18 @@
     {
       "action": "api::event.event.update"
     },
+    {
```
Comment (doolri1276) [inline backend/config/sync/user-role.admin.json:51]: I wonder why there are two of the same announcement apis could you delete one?

File: backend/config/sync/user-role.authenticated.json [hunk 13]
```diff
@@ -18,6 +24,12 @@
     {
       "action": "api::event.event.findOne"
     },
+    {
```
Comment (doolri1276) [inline backend/config/sync/user-role.authenticated.json:27]: I wonder why there are two of the same announcement apis could you delete one?

File: backend/config/sync/user-role.public.json [hunk 14]
```diff
@@ -18,6 +24,12 @@
     {
       "action": "api::event.event.findOne"
     },
+    {
```
Comment (doolri1276) [inline backend/config/sync/user-role.public.json:27]: I wonder why there are two of the same announcement apis could you delete one?

**Review by doolri1276 — CHANGES_REQUESTED** [review]

File: backend/config/sync/user-role.admin.json (hunk 12 above)
Comment (jiahn326) [reply backend/config/sync/user-role.admin.json]: I removed one of them

File: backend/config/sync/user-role.authenticated.json (hunk 13 above)
Comment (jiahn326) [reply backend/config/sync/user-role.authenticated.json]: I removed one of them

File: backend/config/sync/user-role.public.json (hunk 14 above)
Comment (jiahn326) [reply backend/config/sync/user-role.public.json]: I removed one of them

**Review by doolri1276 — APPROVED** [review]
